server/py/services/alerts/crud/alerts.py

```diff
@@ -260,6 +218,61 @@ def populate_event_cache(self, session: sqlalchemy.orm.Session):
         services.alerts.crud.Events().cache_initialized = True
         logger.debug("Finished populating event cache for alerts")
 
+    def should_send_notification(self, alert, state_obj):
+        if alert.criteria:
+            if alert.criteria.period:
+                offset = self._get_event_offset(alert)
+                self._normalize_events(
+                    state_obj,
+                    framework.utils.helpers.string_to_timedelta(
+                        alert.criteria.period, offset, raise_on_error=False
+                    ),
+                )
+            return len(state_obj["events"]) >= alert.criteria.count
+        return True
+
+    def get_number_of_events(self, alert_id: int) -> int:
+        state_obj = self._states.get(alert_id, {"events": []})
+        return len(state_obj["events"])
+
+    @staticmethod
+    def _get_event_offset(alert):
+        if alert.entities.kind == mlrun.common.schemas.alert.EventEntityKind.JOB:
+            return int(mlconfig.monitoring.runs.interval)
+        return 0
+
+    def _handle_notification(self, session, alert, state, state_obj, event_data):
+        update_state = True
+        active = False
+        state["count"] += 1
+        logger.debug("Sending notifications for alert", name=alert.name)
+        AlertNotificationPusher().push(alert, event_data)
+
+        if alert.reset_policy == "auto":
+            self.reset_alert(session, alert.project, alert.name)
+            services.api.crud.AlertActivation().store_alert_activation(
+                session, alert, event_data
+            )
+            update_state = False
+        else:
+            active = True
+            state["active"] = True
+            activation_id = services.api.crud.AlertActivation().store_alert_activation(
+                session, alert, event_data
+            )
+            state_obj["last_activation_id"] = activation_id
+
+        framework.utils.singletons.db.get_db().store_alert_state(
+            session,
+            alert.project,
+            alert.name,
+            count=state["count"],
+            last_updated=event_data.timestamp,
+            obj=state_obj,
+            active=active,
+        )
+        return update_state
```

Comment:
it looks wrong that `update_state` can be false but store the alert state anyway.
I understand that you use it outside for wether to update the state cache, but it's not clear what and why when reading this function.

Reply:
@TomerShor update state indicates whether internal `_states` should be updated... But yeah, looks confusing, maybe rename to keep_cache?